### PR TITLE
Validate Standup Details

### PR DESCRIPTION
**Pull request by SyedSibtainRazvi**
# Related Issue
https://gitlab.com/codemancers/engineering/training/syed-training/-/issues/29

# functionalities Added or Changed

- Added validation for form data at the backend.

# Additional Info
-Adding Express Validator to the Bot to validate Stand-up Details before saving them to the Database

# Checklist
- [ ] Tests
- [ ] Uploaded screenshots

# Screenshots
Validation Unsuccessful
Validation Successful

### CODE REVIEW

File: server/src/routes/standups.router.ts [hunk 1]
```diff
@@ -1,11 +1,35 @@
...
-router.post("/", postStandupHandler);
+router.use(bodyparser.urlencoded({ extended: false }))
+router.use(bodyparser.json())
+
+
+const validateStandup = [
+  body("title").isLength({ min: 5, max: 20 }),
+  body("questions").not().isEmpty(),
+  body("members").not().isEmpty(),
+  body("channel").not().isEmpty(),
```
Comment (SujayPrabhu96) [inline server/src/routes/standups.router.ts:20]: Check the add form in FE and see what other fields have been missed in validation

Comment (SujayPrabhu96) [reply server/src/routes/standups.router.ts:20]: Move these validators to `/helpers/standupValidation.ts`

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/routes/standups.router.ts [hunk 2]
```diff
@@ -1,11 +1,35 @@
...
 } from "../controllers/standup.controller";
 
 const router = express.Router();
-router.post("/", postStandupHandler);
+router.use(bodyparser.urlencoded({ extended: false }))
+router.use(bodyparser.json())
+
+
+const validateStandup = [
+  body("title").isLength({ min: 5, max: 20 }),
```
Comment (SujayPrabhu96) [inline server/src/routes/standups.router.ts:17]: We don't have length validation in FE, so you can remove it and validate if the title is present

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/routes/standups.router.ts (hunk 1 above)
Comment (SyedSibtainRazvi) [reply server/src/routes/standups.router.ts:20]: Done

File: server/src/routes/standups.router.ts (hunk 2 above)
Comment (SyedSibtainRazvi) [reply server/src/routes/standups.router.ts:17]: Changed

File: server/src/helpers/standupValidation.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const { body } = require('express-validator');
+
+export const validateStandup = [
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:3]: Is the prettier installed ?

File: server/src/helpers/standupValidation.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+    body("title").not().isEmpty(),
+    body("questions").not().isEmpty(),
+    body("members").not().isEmpty(),
+    body("channel").not().isEmpty(),
+    body("date").isISO8601().toDate(),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:8]: There is no `date` field sent in FE request. Please check the fields from here: https://github.com/code-mancers/gather-bot/blob/main/server/src/services/standups.service.ts#L27-L32

File: server/src/routes/standups.router.ts [hunk 5]
```diff
@@ -1,11 +1,28 @@
...
-router.post("/", postStandupHandler);
+router.use(bodyparser.urlencoded({ extended: false }))
+router.use(bodyparser.json())
+
+router.post("/", validateStandup, (req: Request, res: Response) => {
+  const errors = validationResult(req)
+  if (!errors.isEmpty()) {
+    return res.status(400).json({ errors: errors.array() });
+  }
+  else {
```
Comment (SujayPrabhu96) [inline server/src/routes/standups.router.ts:21]: Prettier issue

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/helpers/standupValidation.ts (hunk 3 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:3]: Yes

File: server/src/routes/standups.router.ts (hunk 5 above)
Comment (SyedSibtainRazvi) [reply server/src/routes/standups.router.ts:21]: Fixed

File: server/src/helpers/standupValidation.ts (hunk 4 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:8]: Done

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/routes/standups.router.ts [hunk 6]
```diff
@@ -1,11 +1,26 @@
...
 } from "../controllers/standup.controller";
 
 const router = express.Router();
-router.post("/", postStandupHandler);
+router.use(express.urlencoded({ extended: false }));
+router.use(express.json());
+
+router.post("/", validateStandup, (req: Request, res: Response) => {
+  const errors = validationResult(req);
+  if (!errors.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (SyedSibtainRazvi) [reply server/src/routes/standups.router.ts:17]: Understood and corrected

File: server/src/helpers/standupValidation.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+const { body } = require("express-validator");
+
+export const validateStandup = [
+  body("title").not().isEmpty(),
+  body("questions").not().isEmpty(),
+  body("members").not().isEmpty(),
+  body("channel").not().isEmpty(),
+  body("days").isISO8601().toDate(),
```
[the remark this answers is not in the document]
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:8]: Corrected

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/controllers/standup.controller.ts [hunk 8]
```diff
@@ -1,8 +1,9 @@
 import { Request, Response } from "express";
 import { postStandup, getAllStandups } from "../services/standups.service";
 
-export const postStandupHandler = async (req: Request, res: Response) => {
+export const postStandupHandler = async (req: Request, res: Response) => {  
   try {
+    console.log(req);
```
[the remark this answers is not in the document]
Comment (SyedSibtainRazvi) [reply server/src/controllers/standup.controller.ts:6]: Removed

File: server/src/helpers/standupValidation.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+const { body, check } = require("express-validator");
+
+export const validateStandup = [
+  body("title").not().isEmpty(),
+  check("questions.*").not().isEmpty().trim().isLength({ min: 1, max: 10}),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:5]: I have couple of questions
- What does `check("questions.*")` mean ? (Similarly for `members` and `days`)
- We have not set any limit for number of questions. So, do we need to set `isLength({})` ?

> check("days.*").trim().isInt().isLength({ min: 0, max: 6 })
- `days` is an array. So, does `isInt` holds good in this case or do we have any methods like `isArray`

File: server/src/helpers/standupValidation.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+const { body, check } = require("express-validator");
+
+export const validateStandup = [
+  body("title").not().isEmpty(),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:4]: We can also send custom error message:
https://express-validator.github.io/docs/validation-chain-api.html#notempty
https://express-validator.github.io/docs/validation-chain-api.html#notempty

```suggestion
  body("title").trim().notEmpty().withMessage('Title should not be empty'),
```

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/helpers/standupValidation.ts (hunk 10 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:4]: Yes, custom messages can be added

File: server/src/helpers/standupValidation.ts (hunk 9 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:5]: We can use isArray() method as well for days and provide length as well.
Using check("questions.*") is something that I read in a blog while using express-validator, however we can stick to checking in the body.

Length can be removed for the questions as well

File: server/src/helpers/standupValidation.ts [hunk 11]
```diff
@@ -0,0 +1,10 @@
+const { body } = require("express-validator");
+
+export const validateStandup = [
+  body("title").trim().notEmpty().withMessage("Title should not be empty"),
+  body("questions").notEmpty().trim().withMessage("Please enter atleast one question!"),
+  body("members").notEmpty().trim().isLength({ min: 1, max: 20 }),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:6]: There are no limits on members as well

File: server/src/helpers/standupValidation.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+  body("channel").notEmpty().isObject(),
+  body("days").trim().isArray().isLength({ min: 0, max: 6 }),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:8]:
```suggestion
  body("days").trim().isArray().isLength({ min: 1, max: 6 }),
```

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]
Add custom messages for all the fields

File: server/src/helpers/standupValidation.ts (hunk 11 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:6]: Sure, will correct it and add a custom message for all fields.

File: server/src/helpers/standupValidation.ts (hunk 12 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:8]: Corrected

**Comment by SyedSibtainRazvi** [comment]
Validation Unsuccessful

File: server/src/helpers/standupValidation.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
...
+  body("title").trim().notEmpty().withMessage("Title cannot be empty"),
+  body("questions")
+    .notEmpty()
+    .trim()
+    .withMessage("Please enter atleast one question?"),
+  body("members").notEmpty().withMessage("Members cannot be empty"),
+  body("channel")
+    .notEmpty()
+    .isObject()
+    .withMessage("Please enter at least one channel?"),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:13]:
```suggestion
    .withMessage("Please enter at least one channel"),
```

File: server/src/helpers/standupValidation.ts [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  body("days").notEmpty().isArray({min: 1}).withMessage("Please enter a valid day"),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:14]: Is the prettier working ? 🤔

File: server/src/helpers/standupValidation.ts [hunk 15]
```diff
@@ -0,0 +1,16 @@
+const { body } = require("express-validator");
+
+export const validateStandup = [
+  body("title").trim().notEmpty().withMessage("Title cannot be empty"),
+  body("questions")
+    .notEmpty()
+    .trim()
+    .withMessage("Please enter atleast one question?"),
+  body("members").notEmpty().withMessage("Members cannot be empty"),
```
Comment (SujayPrabhu96) [inline server/src/helpers/standupValidation.ts:9]: Can we have `isArray({ min: 1 })` for questions and members ?

**Review by SujayPrabhu96 — CHANGES_REQUESTED** [review]

File: server/src/helpers/standupValidation.ts (hunk 14 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:14]: It works, need to reconfigure it. Some issues

File: server/src/helpers/standupValidation.ts (hunk 13 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:13]: Corrected

File: server/src/helpers/standupValidation.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+const { body } = require("express-validator");
+
+export const validateStandup = [
+  body("title").trim().notEmpty().withMessage("Title cannot be empty"),
+  body("questions")
+    .notEmpty()
+    .trim()
```
[the remark this answers is not in the document]
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:7]: Yes, checked and corrected

File: server/src/helpers/standupValidation.ts [hunk 17]
```diff
@@ -0,0 +1,16 @@
...
+    .withMessage("Please enter atleast one question?"),
```
[the remark this answers is not in the document]
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:8]: Corrected

File: server/src/helpers/standupValidation.ts (hunk 15 above)
Comment (SyedSibtainRazvi) [reply server/src/helpers/standupValidation.ts:9]: I need to check if `isArray` will work on an array of objects, I assume it might now work

**Review by SujayPrabhu96 — APPROVED** [review]

**Pull request merged**